### PR TITLE
Proof of concept - Fix the race with Formatters (issue #375)

### PR DESCRIPTION
**Pull request by haesbaert**
I have no idea where these should be, but this is enough to fix the race, if someone could step in I'd appreciate.

### CODE REVIEW

**Comment by haesbaert** [comment]
issue #375

**Comment by samoht** [comment]
The fix looks ok to me, just two questions:
- what's the cost of looking for something in `DLS` for every print operation? (I suspect this is not really an issue as people are not expected to write to stdout/stderr in a hot loop, but I'm still interested to know the answer :p)
- is there a better way to use both the OCaml 5 API and OCaml 4? For instance, could we have a mock version for DLS somewhere?

File: src/alcotest-engine/dune [hunk 1]
```diff
@@ -15,6 +15,29 @@
...
+ (target safe.ml)
+ (enabled_if
+  (>= ${ocaml_version} 5.0.0))
+ (action
+   (write-file
+     %{target}
+     "let stdout_key = Format.synchronized_formatter_of_out_channel stdout\n
+      let stderr_key = Format.synchronized_formatter_of_out_channel stderr\n
+      let safe_stdout () = Domain.DLS.get stdout_key\n
+      let safe_stderr () = Domain.DLS.get stderr_key\n")))
```
[the remark this answers is not in the document]
Comment (haesbaert) [reply src/alcotest-engine/dune:39]: I think the better, long/term solution would be to have some format5-shims package that hides that away, so people just just that. There is precedence with domain-shims (https://gitlab.com/gasche/domain-shims).
But without some re-engineering in alcotest I don't see a much better way than having to call a function to get the formatter.

**Comment by Sudha247** [comment]
> what's the cost of looking for something in DLS for every print operation?

Lookup on DLS is constant time.

> is there a better way to use both the OCaml 5 API and OCaml 4? For instance, could we have a mock version for DLS somewhere?

We have an OCaml 4 compatible Domain API in [domain-shims](https://gitlab.com/gasche/domain-shims), it also supports DLS -- https://gitlab.com/gasche/domain-shims/-/blob/trunk/lib/domain.ocaml4.mli#L88. The point to note here is domain-shims uses systhreads internally to simulate domains. If that's an issue, a mock DLS implementation for OCaml 4 should be fairly straightforward.

**Comment by haesbaert** [comment]
> The fix looks ok to me, just two questions:
> 
> * what's the cost of looking for something in `DLS` for every print operation? (I suspect this is not really an issue as people are not expected to write to stdout/stderr in a hot loop, but I'm still interested to know the answer :p)

Not that I profiled, but I'd say that it's many magnitudes less than the actual IO operation.

> * is there a better way to use both the OCaml 5 API and OCaml 4? For instance, could we have a mock version for DLS somewhere?

Even if we mock the DLS we would still need to mock the `Format.synchronized_formatter_of_out_channel`.

A simpler alternative could be to convert all prints to some function that grabs a mutex, and then we just use the normal output channel, this way the code could be the same for 4 and 5.

File: src/alcotest-engine/dune (hunk 1 above)
[the remark this answers is not in the document]
Comment (MisterDA) [reply src/alcotest-engine/dune:39]: To be fair, this is what I've also used above to get APIs from OCaml >= 4.12 to iterate over backtraces. Merlin's jump-to-definition doesn't work. The code could be moved to a real file (doesn't change anything for Merlin).

**Review by MisterDA — APPROVED** [review]

**Comment by samoht** [comment]
@haesbaert do you mind adding a change entry? I'm happy to merge/release this quickly to make a release compatible with multi-domains.

**Comment by haesbaert** [comment]
> @haesbaert do you mind adding a change entry? I'm happy to merge/release this quickly to make a release compatible with multi-domains.

I've reworded the commit and added the Changes entry.

File: src/alcotest/alcotest.ml [hunk 2]
```diff
@@ -96,12 +99,12 @@ module Unix_platform (M : Alcotest_engine.Monad.S) = struct
 
   let with_redirect fd_file fn =
     let* () = M.return () in
-    Fmt.(flush stdout) ();
-    Fmt.(flush stderr) ();
+    Fmt.(flush (safe_stdout ())) ();
+    Fmt.(flush (safe_stderr ())) ();
     before_test ~output:fd_file ~stdout ~stderr;
```
Comment (MisterDA) [inline src/alcotest/alcotest.ml:104]: What about the calls to `before_test` and `after_test`? They use a single static variable (per fd) in the C stubs. Seems not so thread-safe.

**Comment by dinosaure** [comment]
Close by #399

**Pull request closed without merging**